### PR TITLE
Add reuse external snapshots scenarios in case blockcommit

### PR DESCRIPTION
**Pull request by chunfuwen**
blockcommit can still succeed after VM destroy and start again

Signed-off-by: chunfuwen <chwen@redhat.com>

### CODE REVIEW

**Comment by chunfuwen** [comment]
Command 'systemctl restart libvirtd.service' finished with 0 after 0.08922028541564941s
Running 'virsh list'
[stdout]  Id    Name                           State
Command 'virsh list' finished with 0 after 0.380237340927124s
[stdout] ----------------------------------------------------
[stdout] 
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.blockcommit.normal_test.single_chain.file_disk.local.reuse-external.notimeout.base.top_inactive

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 1]
```diff
@@ -250,6 +250,34 @@ def check_chain_backing_files(disk_src_file, expect_backing_file=False):
...
+        # Make three external relative path backing files.
+        backing_file_dict = collections.OrderedDict()
+        backing_file_dict["b"] = "b.img"
+        backing_file_dict["c"] = "c.img"
+        backing_file_dict["d"] = "d.img"
+        for key, value in list(backing_file_dict.items()):
+            backing_file_path = os.path.join(root_dir, key)
+            external_snap_shot = "%s/%s" % (backing_file_path, value)
+            snapshot_external_disks.append(external_snap_shot)
+            options = "%s --diskspec vda,file=%s" % (meta_options, external_snap_shot)
```
Comment (chloerh) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:271]: Are we considering the possibility of disk not being 'vda'?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 2]
```diff
@@ -465,6 +499,11 @@ def check_chain_backing_files(disk_src_file, expect_backing_file=False):
             if pivot_opt:
                 blockcommit_options += " --pivot"
 
+        if restart_vm_before_commit:
+            blockcommit_options = "--top vda[%d] --base vda[%d] --verbose --wait --keep-relative" % (1, len(snapshot_external_disks))
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:503]: I'm a little confused about blockcommit_options, it's updated from line 475 to line 500 but will be replaced here. Is it correct?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 3]
```diff
@@ -250,6 +250,34 @@ def check_chain_backing_files(disk_src_file, expect_backing_file=False):
             else:
                 logging.debug("The actual qemu-img output:%s\n", ret)
 
+    def create_reuse_external_snapshots():
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:253]: It overlaps with make_relative_path_backing_files() a little, is it possible to merge them together?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 2 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py]: updated

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 3 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py]: not overlapped, make_relative_path_backing_files() is used to create  backing files, but create_reuse_external_snapshots is used to call virsh create-snapshot-as to bring backing files into action as external snapshots

**Comment by chunfuwen** [comment]
[stdout] 
cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.blockcommit.normal_test.single_chain.block_disk.reuse-external.timeout.shallow.top_active.without_pivot


Command 'ps -o comm 1' finished with 0 after 0.016414165496826172s
[stdout] systemd
legacy start libvirtd daemon IGNORED with function name: restart
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.blockcommit.normal_test.single_chain.block_disk.reuse-external.timeout.shallow.top_active.with_pivot

DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.blockcommit.normal_test.single_chain.block_disk.reuse-external.timeout.shallow.top_inactive

cleaning libvirtd logs...
DATA (filename=output.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stdout.expected) => NOT FOUND (data sources: variant, test, file)
DATA (filename=stderr.expected) => NOT FOUND (data sources: variant, test, file)
PASS 1-type_specific.io-github-autotest-libvirt.virsh.blockcommit.normal_test.single_chain.file_disk.local.reuse-external.notimeout.base.top_inactive

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 1 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py]: updated

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 4]
```diff
@@ -198,14 +205,20 @@ def get_first_disk_source():
...
-        basename = os.path.basename(first_disk_source)
-        root_dir = os.path.dirname(first_disk_source)
+        if pre_set_root_dir is None:
+            first_disk_source = get_first_disk_source()
+            basename = os.path.basename(first_disk_source)
+            root_dir = os.path.dirname(first_disk_source)
+        else:
+            first_disk_source = get_first_disk_source()
+            basename = os.path.basename(first_disk_source)
+            root_dir = pre_set_root_dir
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:221]: Same way to get first_disk_source and basename in 'if' and 'else', would you mind merge them?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 5]
```diff
@@ -215,11 +228,17 @@ def make_relative_path_backing_files():
...
+            backing_file_dict["d"] = "%s/c/c.img" % root_dir
+        disk_format = params.get("disk_format", "qcow2")
         for key, value in list(backing_file_dict.items()):
             backing_file_path = os.path.join(root_dir, key)
-            cmd = ("cd %s && qemu-img create -f qcow2 -o backing_file=%s,backing_fmt=qcow2 %s.img"
-                   % (backing_file_path, value, key))
+            cmd = ("cd %s && qemu-img create -f %s -o backing_file=%s,backing_fmt=%s %s.img"
+                   % (backing_file_path, "qcow2", value, disk_format, key))
             ret = process.run(cmd, shell=True)
+            disk_format = "qcow2"
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:241]: Does it get from params or hard-coded?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py [hunk 6]
```diff
@@ -250,6 +269,46 @@ def check_chain_backing_files(disk_src_file, expect_backing_file=False):
...
+
+    def check_file_in_vm():
+        """
+        Check whether certain image files exists in VM internal.
+        """
+        for img_file in created_image_files_in_vm:
+            status, output = session.cmd_status_output("ls -l /mnt/%s" % img_file)
+            logging.debug(output)
+            if status:
+                test.fail("blockcommit from top to base failed: %s" % output)
```
Comment (Yingshun) [inline libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:310]: Can we report more proper error message here? How about adding some error message like 'No such file '?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 2 above)
Comment (Yingshun) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py]: Okay. Actually, what makes me more confused is that why the option is hard-coded if restart_vm_before_commit. Is it a variable only used to reuse external snapshots' cases?

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 4 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:221]: updated

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 5 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:241]: As per case requirement , the first backend file is raw image, but after all of sequenced backend file format need qcow2, so it is intentional here.

File: libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py (hunk 6 above)
Comment (chunfuwen) [reply libvirt/tests/src/virsh_cmd/domain/virsh_blockcommit.py:310]: updated

**Pull request merged**